crates/fs/private/hamt/node.rs

```diff
@@ -219,45 +222,71 @@ where
             return Ok((Rc::clone(self), None));
         }
 
-        // Otherwise, we go one level deep.
         let value_index = self.get_value_index(bit_index);
         Ok(match &self.pointers[value_index] {
             Pointer::Values(values) => {
                 let mut node = (**self).clone();
                 let value = if values.len() == 1 {
                     // If there is only one value, we can remove the entire pointer.
                     node.bitmask.set(bit_index, false);
-                    let pointer = node.pointers.remove(value_index);
-                    match pointer {
+                    match node.pointers.remove(value_index) {
                         Pointer::Values(mut values) => Some(values.pop().unwrap()),
                         _ => unreachable!(),
                     }
                 } else {
-                    // Otherwise, remove the value.
+                    // Otherwise, remove just the value.
                     let mut values = (*values).clone();
-                    if let Some(i) = values.iter().position(|p| p.key == *key) {
+                    values.iter().position(|p| p.key == *key).map(|i| {
                         let value = values.remove(i);
                         node.pointers[value_index] = Pointer::Values(values);
-                        Some(value)
-                    } else {
-                        None
-                    }
+                        value
+                    })
                 };
 
                 (Rc::new(node), value)
             }
             Pointer::Link(link) => {
                 let child = Rc::clone(link.resolve_value(store).await?);
                 let (child, value) = child.remove_value(hashnibbles, key, store).await?;
+
                 let mut node = (**self).clone();
-                node.pointers[value_index] = Pointer::Link(Link::from(child));
+                if value.is_some() {
+                    // If something has been deleted, we attempt toc canonicalize the pointer.
+                    if let Some(pointer) =
+                        Pointer::Link(Link::from(child)).canonicalize(store).await?
```

Comment:
I was trying to avoid a situation where mutation happens in a call to another function and doesn't happen in the local scope. I probably took the local mutation thing too far 😅. I don't mind making it an `&mut self`.

Reply:
I will merge this PR but I'm going to add the change to the other PR since I have made some other changes there too.